### PR TITLE
Student model file system

### PR DESCRIPTION
**Pull request by JohnHunter809**
- Changed the studentModel file system to save students in individual files in a /students folder
- Added functionality for logging users out in the backend (not implemented from the frontend)

### CODE REVIEW

File: src/main/java/edu/ithaca/dragon/par/ParServer.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import edu.ithaca.dragon.par.pedagogicalModel.TaskGenerator;
 import edu.ithaca.dragon.par.studentModel.StudentModel;
 
+import javax.xml.crypto.Data;
```
Comment (tobydragon) [inline src/main/java/edu/ithaca/dragon/par/ParServer.java]: Why this import?

File: src/main/java/edu/ithaca/dragon/par/ParServer.java [hunk 2]
```diff
@@ -18,22 +19,24 @@
...
 
-    public ParServer(Datastore datastore) throws IOException{
-        this(new QuestionPool(datastore));
-        for (StudentModel studentModel : datastore.loadStudentModels()){
-            studentModelMap.put(studentModel.getUserId(), studentModel);
+    public ImageTask nextImageTask(String userId) throws IOException{
+        StudentModel currentStudent = getOrCreateStudentModel(studentModelMap, userId, datastore);
+        ImageTask imageTask =  TaskGenerator.makeTask(currentStudent);
+
+        //TODO: delete this
```
Comment (tobydragon) [inline src/main/java/edu/ithaca/dragon/par/ParServer.java]: good comment, please do :)

File: src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java [hunk 3]
```diff
@@ -34,24 +37,69 @@ public JsonDatastore(String questionFilePath, String studentModelFilePath) throw
...
     public List<StudentModel> loadStudentModels() throws IOException {
+         //check if filepath exists
         if(studentModelFilePath == null)
             throw new IOException("studentModelFilePath is null");
-        List<StudentModelRecord> studentModelRecords = JsonUtil.listFromJsonFile(studentModelFilePath, StudentModelRecord.class);
+
+        //make a list of studentModelRecords from every file in the filepath
+        List<StudentModelRecord> studentModelRecords = new ArrayList<>();
+        File studentModelFolder = new File(studentModelFilePath);
+        File[] listOfStudentModelFiles = studentModelFolder.listFiles();
```
Comment (tobydragon) [inline src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java]: I think there is someway of checking that this is a directory, we should test what happens if it is given a bad path path to a file (should throw IOException and note the problem).

File: src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java [hunk 4]
```diff
@@ -34,24 +37,69 @@ public JsonDatastore(String questionFilePath, String studentModelFilePath) throw
...
+
+    @Override
+    public void saveStudentModel(StudentModel studentModel) throws IOException{
+        if(studentModelFilePath == null)
+            throw new IOException("studentModelFilePath is null");
+        String fullFilePath = studentModelFilePath + "/" +  studentModel.getUserId() + ".json";
+        JsonUtil.toJsonFile(fullFilePath, new StudentModelRecord(studentModel));
     }
 
+    @Override
```
Comment (tobydragon) [inline src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java]: Is this used? If not, we should leave it out...

File: src/test/resources/students/TestUser100.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (tobydragon) [inline src/test/resources/students/TestUser100.json]: Are these files generated every time? If so, they should be in the autoGenerated folder? Or are they there for examples? In which case they should go under the authoring folder...

File: src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java (hunk 4 above)
Comment (JohnHunter809) [reply src/main/java/edu/ithaca/dragon/par/io/JsonDatastore.java]: This was written for a test, but will probably never be used in the system

File: src/test/resources/students/TestUser100.json (hunk 5 above)
Comment (JohnHunter809) [reply src/test/resources/students/TestUser100.json]: moved to test/resources/author/students/...

**Pull request merged**